usecases/webhooks_usecase.go

```diff
@@ -0,0 +1,224 @@
+	"github.com/checkmarble/marble-backend/models"
+	"github.com/checkmarble/marble-backend/repositories"
+	"github.com/checkmarble/marble-backend/usecases/executor_factory"
+	"github.com/checkmarble/marble-backend/utils"
+	"github.com/google/uuid"
+	"github.com/guregu/null/v5"
+)
+
+type convoyRepository interface {
+	SendWebhookEvent(ctx context.Context, webhook models.Webhook) error
+}
+
+type webhookRepository interface {
+	GetWebhook(ctx context.Context, exec repositories.Executor, webhookId string) (models.Webhook, error)
+	ListWebhooks(ctx context.Context, exec repositories.Executor, filters models.WebhookFilters) ([]models.Webhook, error)
+	CreateWebhook(
+		ctx context.Context,
+		exec repositories.Executor,
+		webhook models.Webhook,
+	) error
+	UpdateWebhook(
+		ctx context.Context,
+		exec repositories.Executor,
+		input models.WebhookUpdate,
+	) error
+}
+
+type enforceSecurityWebhooks interface {
+	CreateWebhook(
+		ctx context.Context,
+		organizationId string,
+		partnerId null.String,
+	) error
+	SendWebhook(
+		ctx context.Context,
+		webhook models.Webhook,
+	) error
+}
+
+type WebhooksUsecase struct {
+	enforceSecurity    enforceSecurityWebhooks
+	executorFactory    executor_factory.ExecutorFactory
+	transactionFactory executor_factory.TransactionFactory
+	convoyRepository   convoyRepository
+	webhookRepository  webhookRepository
+}
+
+func NewWebhooksUsecase(
+	enforceSecurity enforceSecurityWebhooks,
+	executorFactory executor_factory.ExecutorFactory,
+	transactionFactory executor_factory.TransactionFactory,
+	convoyRepository convoyRepository,
+	webhookRepository webhookRepository,
+) WebhooksUsecase {
+	return WebhooksUsecase{
+		enforceSecurity:    enforceSecurity,
+		executorFactory:    executorFactory,
+		transactionFactory: transactionFactory,
+		convoyRepository:   convoyRepository,
+		webhookRepository:  webhookRepository,
+	}
+}
+
+func (usecase WebhooksUsecase) CreateWebhook(
+	ctx context.Context,
+	input models.WebhookCreate,
+) error {
+	err := usecase.enforceSecurity.CreateWebhook(ctx, input.OrganizationId, input.PartnerId)
+	if err != nil {
+		return err
+	}
+
+	webhook, err := executor_factory.TransactionReturnValue(
+		ctx,
+		usecase.transactionFactory,
+		func(tx repositories.Executor) (models.Webhook, error) {
+			webhookId := uuid.New().String()
+			now := time.Now()
+			err = usecase.webhookRepository.CreateWebhook(ctx,
+				usecase.executorFactory.NewExecutor(), models.Webhook{
+					Id:               webhookId,
+					CreatedAt:        now,
+					UpdatedAt:        now,
+					SendAttemptCount: 0,
+					DeliveryStatus:   models.Scheduled,
+					OrganizationId:   input.OrganizationId,
+					PartnerId:        input.PartnerId,
+					EventType:        input.EventType,
+					EventData:        input.EventData,
+				})
+			if err != nil {
+				return models.Webhook{}, fmt.Errorf("error while creating webhook: %w", err)
+			}
+			return usecase.webhookRepository.GetWebhook(ctx, tx, webhookId)
+		})
+
+	go func() {
```

Comment:
Ok I'm thinking this through again.
We would like to insert the webhook in the same transaction in which we do the insert/update.
Once it's committed (and not before), we want to send the webhook to convoy, doing any necessary bookkeeping of the webhook stored on our side. So in practice it's not clear to me that we can or should call `sendWebhooks` from within this function - as the transaction would not yet be committed.